sync-r2.sh

```diff
@@ -0,0 +1,99 @@
+# locally. Files in the root directory with names starting with a dot or an
+# underscore are not uploaded. After the upload the Cloudflare CDN cache is
+# purged for the affected files.
+#
+# The script assumes that the AWS CLI tool is installed and already configured
+# with credentials allowing it to modify the bucket.
+#
+# NOTE: There's no built-in mechanism for updating an S3 bucket in an atomic
+# way. Only individual file updates are atomic. This means that during the
+# sync clients will see the intermediate state with some files missing or not
+# yet updated. Since the binaries are never modified or removed from the repository
+# under normal circumstances, updating file lists last is enough to alleviate this.
+#
+# When running multiple instances of this script concurrently on different
+# revisions it's theoretically possible to end up with any combination of
+# their files in the bucket so it should be avoided.
+#
+# WARNING: The script destructively modifies the working copy. Always run it
+# on a fresh clone!
+#------------------------------------------------------------------------------
+
+set -eo pipefail
+
+die() { >&2 echo "ERROR: $@" && false; }
+
+r2_account_id="$1"
+r2_zone="$2"
+r2_bucket_name="$3"
+cloudflare_zone_id="$4"
+cloudflare_cache_host="$5"
+cloudflare_api_token="$6"
+
+(( $# == 6 )) || die "Expected exactly 6 parameters."
+
+r2_endpoint="https://${r2_account_id}.r2.cloudflarestorage.com"
+r2_bucket_uri="s3://${r2_bucket_name}"
+
+echo "===> Using Cloudflare R2 bucket '${r2_bucket_name}' (zone: ${r2_zone}) via account ${r2_account_id}"
+
+[[ $(git rev-parse --is-shallow-repository) == false ]] || die "This script requires access to full git history to be able to set file timestamps correctly."
+
+echo "===> Updating file modification timestamps to match commits"
+# NOTE: `aws s3 sync` compares file timestamp and size to decide whether to upload it or not.
+readarray -t files < <(git ls-files)
+for file in "${files[@]}"; do
+    full_time="$(git log --max-count 1 --pretty=format:%cd --date=iso -- "$file")"
+    unix_timestamp="$(date --date="$full_time" +%Y%m%d%H%M.%S)"
+    touch -m -t "$unix_timestamp" "$file"
+done
+
+echo "===> Removing files that should not be uploaded to the bucket"
+# NOTE: This ensures that they will be deleted from the bucket if they're already there.
+# If we used `aws s3 sync --delete --exclude` instead, they would not get deleted.
+find . -path './.*' -delete
+find . -path './_*' -delete
+
+echo "===> Adding compatibility symlinks for files containing plus signs in the name"
+# NOTE: This is a quick'n'dirty workaround for Amazon S3 decoding plus sign in paths
+# as a space even though this substitution is only supposed to happen in a query string.
+# See https://forums.aws.amazon.com/thread.jspa?threadID=55746
+find . \
+    -regex "^\(.*/\)*[^/]*\+[^/]*$" \
+    -exec bash -c 'ln --symbolic --no-target-directory "$(basename "{}")" "$(dirname "{}")/$(basename "{}" | tr "+" " ")"' \;
```

Comment:
I wonder if R2 actually needs this workaround. S3 did when I wrote the script, but it's possible that Cloudflare did not copy the API bug for bug :)

All binaries have `+` in their names, so being able to drop it would halve our storage use and make the sync faster. So perhaps worth checking. Though in the end the thing we pay the most for is bandwidth so it won't affect cost much.

If you decide to leave it as is, without checking, I'd at least add a comment that we have not checked if it's still needed for Cloudflare.